### PR TITLE
syncer: replace `{schema, table}` with `filter.Table` Part 1

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve? 

Part of https://github.com/pingcap/dm/issues/2078 and https://github.com/pingcap/dm/issues/2064

### What is changed and how it works?

- Add a alias `Table` for `filter.Table`.
- Replace `{schema, table}` with `Table

### Check List 

Tests 

Code changes

 - Has interface methods change

### CODE REVIEW

File: syncer/checkpoint.go [hunk 1]
```diff
@@ -103,7 +103,7 @@ func (b *binlogPoint) flush() {
 	b.flushedTI = b.ti
 }
 
-func (b *binlogPoint) rollback(schemaTracker *schema.Tracker, schema string) (isSchemaChanged bool) {
```
Comment (lance6716) [inline syncer/checkpoint.go:106]: I like the old name which is "schema.tracker" 😂 so if we use `filter.Table` we can keep it.

Besides, schema package only deals about upstream table structure related things.

Comment (okJiang) [reply syncer/checkpoint.go:106]: Is it reasonable in semantically that using `filter.Table`?🤔

File: syncer/ddl.go [hunk 2]
```diff
@@ -262,21 +263,21 @@ func (s *Syncer) dropSchemaInSharding(tctx *tcontext.Context, sourceSchema strin
...
-		targetSchema, targetTable := utils.UnpackTableID(name)
+		targetTable := utils.UnpackTableID(name)
 		sourceIDs := make([]string, 0, len(tables))
 		for _, table := range tables {
-			sourceID, _ := utils.GenTableID(table[0], table[1])
+			sourceID := utils.GenTableID(table)
 			sourceIDs = append(sourceIDs, sourceID)
 		}
-		err := s.sgk.LeaveGroup(targetSchema, targetTable, sourceIDs)
+		err := s.sgk.LeaveGroup(targetTable, sourceIDs)
```
[the remark this answers is not in the document]
Comment (Ehco1996) [reply syncer/ddl.go:280]: the way we called it before was too confusing，i found: 
* tableID,
* schemaId,
* sourceID ,
* newID 
but they all come from `GenTableID` 😂😂😂,maybe you can unify these names in this PR

File: pkg/utils/common.go [hunk 3]
```diff
@@ -248,19 +248,29 @@ func NonRepeatStringsEqual(a, b []string) bool {
 }
 
 // GenTableID generates table ID.
-func GenTableID(schema, table string) (id string, isSchemaOnly bool) {
-	if len(table) == 0 {
-		return "`" + schema + "`", true
-	}
-	return "`" + schema + "`.`" + table + "`", false
+func GenTableID(table *filter.Table) string {
+	return table.String()
```
Comment (Ehco1996) [inline pkg/utils/common.go:252]: what if table.Name is empty🤔

Comment (okJiang) [reply pkg/utils/common.go:252]: `filter.table` in `tidb-tools` is same:

```go
// String implements the fmt.Stringer interface.
func (t *Table) String() string {
	if len(t.Name) > 0 {
		return fmt.Sprintf("`%s`.`%s`", t.Schema, t.Name)
	}
	return fmt.Sprintf("`%s`", t.Schema)
}
```

> https://github.com/pingcap/tidb-tools/blob/288ab02f1c79faaff0098bfe66d1d9da48607d23/pkg/table-filter/compat.go#L35-L41

File: syncer/ddl.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (okJiang) [reply syncer/ddl.go:280]: SGTM. We can distinguish source/target and schema/table in naming.

File: syncer/syncer.go [hunk 4]
```diff
@@ -2122,7 +2122,7 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
 			ec.tctx.L().Info("re-replicate shard group was completed", zap.String("event", "row"), zap.Stringer("re-shard", ec.shardingReSync))
 			return ec.closeShardingResync()
 		}
-		if ec.shardingReSync.targetSchema != schemaName || ec.shardingReSync.targetTable != tableName {
+		if !reflect.DeepEqual(ec.shardingReSync.targetTable, &filter.Table{Schema: schemaName, Name: tableName}) {
```
Comment (Ehco1996) [inline syncer/syncer.go:2125]: perfer not to use reflect here for better performance, maybe use `ec.shardingReSync.targetTable.String()==&filter.Table{Schema: schemaName, Name: tableName}.String()`

**Review by Ehco1996 — COMMENTED** [review]
please remeber fix lint 
rest LGTM

File: syncer/ddl.go [hunk 5]
```diff
@@ -262,21 +262,21 @@ func (s *Syncer) dropSchemaInSharding(tctx *tcontext.Context, sourceSchema strin
...
 	for name, tables := range sources {
-		targetSchema, targetTable := utils.UnpackTableID(name)
-		sourceIDs := make([]string, 0, len(tables))
+		targetTable := utils.UnpackTableID(name)
+		sourceTableIDs := make([]string, 0, len(tables))
 		for _, table := range tables {
-			sourceID, _ := utils.GenTableID(table[0], table[1])
-			sourceIDs = append(sourceIDs, sourceID)
+			sourceTableID := utils.GenTableID(table)
+			sourceTableIDs = append(sourceTableIDs, sourceTableID)
```
Comment (lance6716) [inline syncer/ddl.go:277]: seems can be merged into one line, because `GenTableID` returns one value now

File: syncer/sharding_group.go [hunk 6]
```diff
@@ -277,18 +278,18 @@ func (sg *ShardingGroup) Sources() map[string]bool {
...
-func (sg *ShardingGroup) Tables() [][]string {
+func (sg *ShardingGroup) Tables() []*filter.Table {
 	sources := sg.Sources()
-	tables := make([][]string, 0, len(sources))
+	tables := make([]*filter.Table, 0, len(sources))
 	for id := range sources {
-		schema, table := utils.UnpackTableID(id)
-		tables = append(tables, []string{schema, table})
+		table := utils.UnpackTableID(id)
+		tables = append(tables, table)
```
Comment (lance6716) [inline syncer/sharding_group.go:286]: ditto

File: syncer/sharding_group.go [hunk 7]
```diff
@@ -298,10 +299,10 @@ func (sg *ShardingGroup) UnresolvedTables() [][]string {
 		return nil
 	}
 
-	tables := make([][]string, 0, len(sg.sources))
+	tables := make([]*filter.Table, 0, len(sg.sources))
 	for id := range sg.sources {
-		schema, table := utils.UnpackTableID(id)
-		tables = append(tables, []string{schema, table})
+		table := utils.UnpackTableID(id)
+		tables = append(tables, table)
```
Comment (lance6716) [inline syncer/sharding_group.go:305]: ditto

File: syncer/sharding_group.go [hunk 8]
```diff
@@ -611,24 +614,22 @@ func (k *ShardingGroupKeeper) UnresolvedGroups() []*pb.ShardingGroup {
...
 }
 
 // ActiveDDLFirstLocation returns the binlog position of active DDL.
-func (k *ShardingGroupKeeper) ActiveDDLFirstLocation(targetSchema, targetTable string) (binlog.Location, error) {
-	group := k.Group(targetSchema, targetTable)
+func (k *ShardingGroupKeeper) ActiveDDLFirstLocation(targetTable *filter.Table) (binlog.Location, error) {
 	k.Lock()
 	defer k.Unlock()
-	if group != nil {
+	if group := k.Group(targetTable); group != nil {
```
Comment (lance6716) [inline syncer/sharding_group.go:628]: `k.Group` will call `k.RLock()`, but golang's lock is not reentrant

File: pkg/terror/error_list.go [hunk 9]
```diff
@@ -982,7 +982,7 @@ var (
 	ErrSyncUnitDDLWrongSequence          = New(codeSyncUnitDDLWrongSequence, ClassSyncUnit, ScopeInternal, LevelHigh, "detect inconsistent DDL sequence from source %+v, right DDL sequence should be %+v", "Please use `show-ddl-locks` command for more details.")
 	ErrSyncUnitDDLActiveIndexLarger      = New(codeSyncUnitDDLActiveIndexLarger, ClassSyncUnit, ScopeInternal, LevelHigh, "activeIdx %d larger than length of global DDLItems: %v", "")
 	ErrSyncUnitDupTableGroup             = New(codeSyncUnitDupTableGroup, ClassSyncUnit, ScopeInternal, LevelHigh, "table group %s exists", "")
-	ErrSyncUnitShardingGroupNotFound     = New(codeSyncUnitShardingGroupNotFound, ClassSyncUnit, ScopeInternal, LevelHigh, "sharding group for `%s`.`%s` not found", "")
+	ErrSyncUnitShardingGroupNotFound     = New(codeSyncUnitShardingGroupNotFound, ClassSyncUnit, ScopeInternal, LevelHigh, "sharding group for %s not found", "")
```
Comment (lance6716) [inline pkg/terror/error_list.go:985]:
```suggestion
	ErrSyncUnitShardingGroupNotFound     = New(codeSyncUnitShardingGroupNotFound, ClassSyncUnit, ScopeInternal, LevelHigh, "sharding group for %v not found", "")
```

File: syncer/syncer.go [hunk 10]
```diff
@@ -530,8 +530,8 @@ func (s *Syncer) initShardingGroups(ctx context.Context, needCheck bool) error {
 	// add sharding group
 	for targetSchema, mSchema := range mapper {
 		for targetTable, sourceIDs := range mSchema {
-			tableID, _ := utils.GenTableID(targetSchema, targetTable)
-			_, _, _, _, err := s.sgk.AddGroup(targetSchema, targetTable, sourceIDs, loadMeta[tableID], false)
+			targetTableID := utils.GenTableID(&filter.Table{Schema: targetSchema, Name: targetTable})
```
Comment (lance6716) [inline syncer/syncer.go:533]: we can reuse a `filter.Table`

File: syncer/syncer.go [hunk 11]
```diff
@@ -2162,11 +2162,10 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
 	}
 
 	if s.cfg.ShardMode == config.ShardPessimistic {
-		source, _ := utils.GenTableID(originSchema, originTable)
-		if s.sgk.InSyncing(schemaName, tableName, source, *ec.currentLocation) {
+		if s.sgk.InSyncing(&filter.Table{Schema: originSchema, Name: originTable}, &filter.Table{Schema: schemaName, Name: tableName}, *ec.currentLocation) {
 			// if in unsync stage and not before active DDL, ignore it
 			// if in sharding re-sync stage and not before active DDL (the next DDL to be synced), ignore it
-			ec.tctx.L().Debug("replicate sharding DDL, ignore Rows event", zap.String("event", "row"), zap.String("source", source), log.WrapStringerField("location", ec.currentLocation))
+			ec.tctx.L().Debug("replicate sharding DDL, ignore Rows event", zap.String("event", "row"), log.WrapStringerField("location", ec.currentLocation))
```
Comment (lance6716) [inline syncer/syncer.go:2168]: lost a zap field

**Review by lance6716 — COMMENTED** [review]
rest lgtm

**Review by lance6716 — APPROVED** [review]
can use `// nolint:ifshort` to suppress lint

File: syncer/syncer.go [hunk 12]
```diff
@@ -2192,11 +2193,13 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
...
 	if s.cfg.ShardMode == config.ShardPessimistic {
-		source, _ := utils.GenTableID(originSchema, originTable)
-		if s.sgk.InSyncing(schemaName, tableName, source, *ec.currentLocation) {
+		if s.sgk.InSyncing(&filter.Table{Schema: originSchema, Name: originTable}, &filter.Table{Schema: schemaName, Name: tableName}, *ec.currentLocation) {
 			// if in unsync stage and not before active DDL, ignore it
 			// if in sharding re-sync stage and not before active DDL (the next DDL to be synced), ignore it
-			ec.tctx.L().Debug("replicate sharding DDL, ignore Rows event", zap.String("event", "row"), zap.String("source", source), log.WrapStringerField("location", ec.currentLocation))
+			ec.tctx.L().Debug("replicate sharding DDL, ignore Rows event",
+				zap.String("event", "row"),
+				zap.Any("source", &filter.Table{Schema: originSchema, Name: originTable}),
```
Comment (lichunzhu) [inline syncer/syncer.go:2201]:
```suggestion
				zap.Stringer("source", &filter.Table{Schema: originSchema, Name: originTable}),
```

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**